### ISSUE TITLE
Regression from stable: "Unexpected tail in unsized_info_ty"

### ISSUE THREAD

**Issue by withoutboats** [rust-lang/rust#42210]
There's been a regression in a recent nightly regarding dispatch on unsized types:

```rust
trait Foo {
    fn foo() { }
}

struct Bar;

trait Baz {
}

impl Foo for (Bar, Baz) { }


fn main() {
    <(Bar, Baz) as Foo>::foo()
}
```

https://is.gd/NjyQaO

**Comment by withoutboats** [comment]
It looks like this function (`type_of::unsized_info_ty`) just doesn't handle unsized tuples. Can't tell how it started getting called but it did.

**Comment by est31** [comment]
Using the [bisect-rust](https://github.com/Mark-Simulacrum/bisect-rust) tool I've bisected the regression to commit 9fa25a7df3a1101a1dac611f3e45fa5b09836d27, PR #42015

cc @nikomatsakis

**Comment by nikomatsakis** [comment]
Huh, interesting. I sort of expect this program to be an error, because (IIRC) in tuples we do not permit unsized types, but maybe that's not true or I'm misremembering.

**Comment by withoutboats** [comment]
We permit the final member of a tuple to be unsized: rust-lang/rfcs/pull/1592

**Comment by nikomatsakis** [comment]
ok, that makes sense

**Comment by nikomatsakis** [comment]
triage: P-high

I'll investigate and try to stay on top of this. Oddly, I'm not reproducing on my local build, though it includes a few further changes.

**Comment by Mark-Simulacrum** [comment]
I am unable to reproduce as well in the same nightly as playpen uses on both macOS and Ubuntu, so there is something interesting going on here for sure I think.

**Comment by est31** [comment]
Confirmed. The ICE is reproducible in `rustc 1.19.0-nightly (28fd1e519 2017-05-27)` but not reproducible any more in `rustc 1.19.0-nightly (d47cf08d5 2017-05-28)`.

Should we add a test and close this, or should we investigate?

**Comment by arielb1** [comment]
Niko fixed this in #42251 but put the wrong issue on the fixes line. Closing.

**Comment by Mark-Simulacrum** [comment]
Huh, I wasn't able to reproduce with the 05-27 nightly, but that may have been a fluke of my environment; not sure. I'm fine with closing of course, though.
